src/instructlab/model/torchrun_train.py

```diff
@@ -0,0 +1,304 @@
+            / f"samples_{(global_step*8)}"
+        )
+
+        print(f"Saving Model to: {output_dir}")
+        model_state = model.state_dict()
+        output_dir.mkdir(parents=True, exist_ok=True)
+        output_model_file = output_dir / "pytorch_model.bin"
+        output_config_file = output_dir / "config.json"
+
+        tmp_conf = copy(model.config)
+
+        torch.save(model_state, str(output_model_file))
+        tmp_conf.to_json_file(str(output_config_file))
+        tokenizer.save_pretrained(str(output_dir))
+
+        llamacpp_convert_to_gguf.convert_llama_to_gguf(
+            model=output_dir,
+            pad_vocab=True,
+            skip_unknown=True,
+            outfile=os.path.join(output_dir, "pytorch_model.gguf"),
+        )
+
+
+def pad_collate_fn(batch, pad_token_id):
+    lens = np.array([len(item["input_ids"]) for item in batch])
+    max_len = max(lens)
+
+    input_ids = torch.stack(
+        [
+            F.pad(
+                item["input_ids"],
+                (max_len - len(item["input_ids"]), 0),
+                mode="constant",
+                value=pad_token_id,
+            )
+            for item in batch
+        ]
+    )
+    labels = torch.stack(
+        [
+            F.pad(
+                item["labels"],
+                (max_len - len(item["labels"]), 0),
+                mode="constant",
+                value=-100,
+            )
+            for item in batch
+        ]
+    )
+    num_loss_counted_tokens = (labels != -100).sum()
+
+    attention_mask = torch.stack(
+        [
+            F.pad(
+                item["attention_mask"],
+                (max_len - len(item["attention_mask"]), 0),
+                mode="constant",
+                value=0,
+            )
+            for item in batch
+        ]
+    )
+    print(
+        f"\033[96m total tokens: {max_len * len(batch)} num samples: {len(batch)} num padding tokens: {max_len * len(batch) - lens.sum()} - rank: {0} "
+        f"max len: {max_len} min len: {min(lens)} avg len: {lens.mean()} "
+        f"num_loss_counted_tokens: {num_loss_counted_tokens}\033[0m"
+    )
+
+    return {
+        "input_ids": input_ids,
+        "labels": labels,
+        "num_loss_counted_tokens": num_loss_counted_tokens,
+        "attention_mask": attention_mask,
+    }
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--chat-tmpl-path", type=str)
+    parser.add_argument("--model-path", type=str)
+    parser.add_argument("--effective-batch-size", type=int)
+    parser.add_argument("--max-batch-len", type=int)
+    parser.add_argument("--data-output-dir", type=str)
+    parser.add_argument("--is-padding-free", action="store_true")
+    parser.add_argument("--data-path", type=str)
+    parser.add_argument("--max-seq-len", type=int)
+    parser.add_argument("--device", type=str, default="cpu")
+    parser.add_argument("--num-epochs", type=int)
+    args = parser.parse_args()
+    main(args)
```

Comment:
Shouldn't this be using `click` and not `argparse`?

Reply:
removed, I realized instead of using torchrun we could just call the func directly since we do not want to support distributed training with this version